services/brig/src/Brig/Whitelist.hs

```diff
@@ -1,44 +1,68 @@
 -- | > docs/reference/user/activation.md {#RefActivationWhitelist}
 --
 -- Email/phone whitelist.
-module Brig.Whitelist (Whitelist (..), verify) where
+module Brig.Whitelist
+    ( Whitelist (..)
+    , InternalWhitelist (..)
+    , verifyInternal
+    , verifyService
+    ) where
 
 import Imports
 import Data.Aeson
+import Data.Aeson.TH (deriveJSON)
 import Bilge.IO
 import Bilge.Request
 import Bilge.Response
 import Bilge.Retry
 import Brig.Types
 import Control.Monad.Catch (MonadMask, throwM)
 import Control.Retry
-import Data.Text
+import qualified Data.Text as T
 import Data.Text.Encoding (encodeUtf8)
 import Network.HTTP.Client (HttpException (..), HttpExceptionContent (..), parseRequest)
+import Util.Options.Common (toOptionFieldName)
 
 -- | A service providing a whitelist of allowed email addresses and phone numbers
+-- DEPRECATED: use internal whitelisting instead!
 data Whitelist = Whitelist
     { whitelistUrl  :: !Text     -- ^ Service URL
-    , whitelistUser :: !Text     -- ^ Username
-    , whitelistPass :: !Text     -- ^ Password
+    , whitelistUser :: !Text     -- ^ Service Username (basic auth)
+    , whitelistPass :: !Text     -- ^ Service Password (basic auth)
+
     } deriving (Show, Generic)
 
 instance FromJSON Whitelist
 
+data InternalWhitelist = InternalWhitelist
+    { internalWhitelistDomains :: ![Text]} deriving (Show, Generic)
+
+deriveJSON toOptionFieldName 'internalWhitelistDomains
+
+-- | Check internal whitelist for given email/phone number. Currently this
+-- consults a statically configured whitelist of allowed email domains..
+verifyInternal :: (MonadIO m, MonadMask m, MonadHttp m) => InternalWhitelist -> Either Email Phone -> m Bool
```

Comment:
Why not simply make this function signature `verifyInternal :: InternalWhitelist -> Either Email Phone -> Bool` ? You're not using monadic effects here.

This would also make unit testing this function easy.

Reply:
To keep the API symmetrical between `verifyInternal` and `verifyService`. In addition, `verifyInternal` will need to be in IO (or something) once it starts checking cassandra for specific whitelisting.

Reply:
Not a good reason (You can always `pure Whitelist.verifyInternal b key` in Handler.hs.), and could be refactored later if the need arises. Also, it will never need MonadHttp, nor MonadMask, even if it talks to cassandra (it would need `MonadClient` at that point). I would generally recommend to keep function signatures as lean as you can.

Reply:
Done.